File: modules/openlmis-web/src/main/webapp/public/js/requisition-group/controller/requisition-group-controller.js
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 *  This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 *  You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
 */

function RequisitionGroupController($scope, requisitionGroupData, messageService, $location, RequisitionGroups, SupervisoryNodesSearch, programs, schedules) {

  var currentProgramSchedule, programScheduleUnderEdit;
  $scope.$parent.message = "";

  if (requisitionGroupData) {
    $scope.requisitionGroup = requisitionGroupData.requisitionGroup;
    $scope.requisitionGroupMembers = requisitionGroupData.requisitionGroupMembers;
    $scope.requisitionGroupProgramSchedules = requisitionGroupData.requisitionGroupProgramSchedules;
    _.each($scope.requisitionGroupProgramSchedules, function (requisitionGroupProgramSchedule) {
      requisitionGroupProgramSchedule.underEdit = false;
    });
  }
  else {
    $scope.requisitionGroup = {};
    $scope.requisitionGroupMembers = [];
    $scope.requisitionGroupProgramSchedules = [];
  }

  $scope.schedules = schedules;
  $scope.newProgramSchedule = {};

  refreshAndSortPrograms();

  function refreshAndSortPrograms() {
    var selectedProgramIds = _.pluck(_.pluck($scope.requisitionGroupProgramSchedules, 'program'), 'id');
    $scope.programs = _.reject(programs, function (program) {
      return _.contains(selectedProgramIds, program.id);
    });

    $scope.programs = _.sortBy($scope.programs, function (program) {
      return program.name.toLowerCase();
    });
    $scope.programMessage = $scope.programs.length ? "label.select.program" : "label.noProgramLeft";
  }

  $scope.addNewRow = function () {
    $scope.addNew = true;
  };

  $scope.updateSchedule = function (index) {
    $scope.requisitionGroupProgramSchedules[index].processingSchedule = _.find($scope.schedules, function (schedule) {
      return $scope.requisitionGroupProgramSchedules[index].processingSchedule.id == schedule.id;
    });
  };

  $scope.edit = function (index) {
    currentProgramSchedule = angular.copy($scope.requisitionGroupProgramSchedules[index]);
    $scope.requisitionGroupProgramSchedules[index].underEdit = true;
  };

  $scope.clearDropOffFacility = function (index) {
    $scope.requisitionGroupProgramSchedules[index].dropOffFacility = undefined;
  };

  $scope.cancelEdit = function (index) {
    $scope.requisitionGroupProgramSchedules[index] = currentProgramSchedule;
    $scope.requisitionGroupProgramSchedules[index].underEdit = false;
    currentProgramSchedule = undefined;
    closeSlider();
  };

  $scope.cancelAdd = function () {
    $scope.newProgramSchedule = {};
    $scope.addNew = false;
    closeSlider();
  };

  $scope.findProgramScheduleUnderEdit = function () {
    return _.find($scope.requisitionGroupProgramSchedules, function (programSchedule) {
      return programSchedule.underEdit === true;
    });
  };

  $scope.saveEditableRow = function (index) {
    $scope.requisitionGroupProgramSchedules[index].underEdit = false;
    programScheduleUnderEdit = $scope.findProgramScheduleUnderEdit();
    $scope.requisitionGroupProgramSchedules[index].dropOffFacility = (programScheduleUnderEdit && programScheduleUnderEdit.dropOffFacility) || $scope.requisitionGroupProgramSchedules[index].dropOffFacility;
    closeSlider();
  };

  $scope.cancel = function () {
    $scope.$parent.successMessage = "";
    $scope.$parent.requisitionGroupId = undefined;
    $location.path('#/search');
  };

  loadSupervisoryNode();

  $scope.addProgramSchedules = function () {
    $scope.requisitionGroupProgramSchedules.push($scope.newProgramSchedule);
    refreshAndSortPrograms();
    $scope.newProgramSchedule = {};
    $scope.addNew = false;
    closeSlider();
  };

  $scope.remove = function (programId) {
    $scope.requisitionGroupProgramSchedules = _.reject($scope.requisitionGroupProgramSchedules, function (programSchedule) {
      return programSchedule.program.id === programId;
    });
    refreshAndSortPrograms();
  };

  $scope.save = function () {
    if ($scope.requisitionGroupForm.$error.required || !$scope.requisitionGroup.supervisoryNode) {
      $scope.showError = "true";
      $scope.error = 'form.error';
      $scope.successMessage = "";
      return;
    }

    if ($scope.findProgramScheduleUnderEdit()) {
      $scope.error = 'error.program.schedules.not.done';
      return;
    }

    if ($scope.requisitionGroup.id) {
      RequisitionGroups.update({id: $scope.requisitionGroup.id},
        {"requisitionGroup": $scope.requisitionGroup, "requisitionGroupMembers": $scope.requisitionGroupMembers, "requisitionGroupProgramSchedules": $scope.requisitionGroupProgramSchedules},
        success, error);
    }
    else {
      RequisitionGroups.save({},
        {"requisitionGroup": $scope.requisitionGroup, "requisitionGroupMembers": $scope.requisitionGroupMembers, "requisitionGroupProgramSchedules": $scope.requisitionGroupProgramSchedules},
        success, error);
    }
  };

  $scope.toggleSlider = function (index) {
    $scope.currentSlider = $scope.showSlider ? undefined : index;
    $scope.showSlider = !$scope.showSlider;
    $scope.extraParams = {"virtualFacility": null, "enabled": true};
  };

  var closeSlider = function () {
    if ($scope.showSlider === true)
      $scope.toggleSlider();
  };

  $scope.associate = function (facility) {
    programScheduleUnderEdit = $scope.findProgramScheduleUnderEdit();
    if (programScheduleUnderEdit) {
      programScheduleUnderEdit.dropOffFacility = facility;
    }
    else {
      $scope.newProgramSchedule.dropOffFacility = facility;
    }
    closeSlider();
  };

  $scope.toggleMultipleFacilitiesSlider = function () {
    $scope.showMultipleFacilitiesSlider = !$scope.showMultipleFacilitiesSlider;
    $scope.extraMultipleParams = {"virtualFacility": null, "enabled": true};
  };

  $scope.addMembers = function (tempFacilities) {
    var duplicateMembers = _.intersection(_.pluck(_.pluck($scope.requisitionGroupMembers, 'facility'), 'id'),
      _.pluck(tempFacilities, "id"));
    if (duplicateMembers.length > 0) {
      $scope.duplicateFacilityName = _.find($scope.requisitionGroupMembers, function (member) {
        return member.facility.id == duplicateMembers[0];
      }).facility.name;
      $scope.message = undefined;
      return false;
    }

    $.each(tempFacilities, function (index, tempFacility) {
      var newMember = {"facility": tempFacility, "requisitionGroup": $scope.requisitionGroup};
      $scope.requisitionGroupMembers.push(newMember);
    });

    $scope.requisitionGroupMembers = _.sortBy($scope.requisitionGroupMembers, function (member) {
      return member.facility.code.toLowerCase();
    });

    $scope.showMultipleFacilitiesSlider = !$scope.showMultipleFacilitiesSlider;
    $scope.message = messageService.get('facilities.added.successfully');
    $scope.duplicateFacilityName = undefined;
    return true;
  };

  $scope.$watch('showMultipleFacilitiesSlider', function () {
    $scope.duplicateFacilityName = undefined;
  });

  $scope.removeMember = function (memberFacilityId) {
    $scope.requisitionGroupMembers = _.filter($scope.requisitionGroupMembers, function (member) {
      return member.facility.id != memberFacilityId;
    });
  };

  var success = function (data) {
    $scope.error = "";
    $scope.$parent.successMessage = data.success;
    $scope.$parent.requisitionGroupId = data.requisitionGroupId;
    $scope.showError = false;
    $location.path('#/search');
  };

  var error = function (data) {
    $scope.$parent.successMessage = "";
    $scope.error = data.data.error;
    $scope.showError = true;
  };

  var compareQuery = function () {
    if (!isUndefined($scope.previousQuery)) {
      return $scope.query.substr(0, 3) !== $scope.previousQuery.substr(0, 3);
    }
    return true;
  };

  $scope.showSupervisoryNodeSearchResults = function () {
    if ($scope.query === undefined || $scope.query.length < 3) return;

    if (compareQuery()) {
      SupervisoryNodesSearch.get({searchParam: $scope.query}, function (data) {
        $scope.supervisoryNodes = data.supervisoryNodeList;
        $scope.filteredNodeList = $scope.supervisoryNodes;
        $scope.previousQuery = $scope.query;
        $scope.nodeResultCount = $scope.filteredNodeList.length;
      }, {});
    }
    else {
      $scope.filteredNodeList = _.filter($scope.supervisoryNodes, function (node) {
        return node.name.toLowerCase().indexOf($scope.query.toLowerCase()) !== -1;
      });
      $scope.nodeResultCount = $scope.filteredNodeList.length;
    }
  };

  $scope.setSelectedSupervisoryNode = function (node) {
    $scope.requisitionGroup.supervisoryNode = node;
    $scope.nodeSelected = node;
    loadSupervisoryNode();
    $scope.query = undefined;
  };

  $scope.clearSelectedNode = function () {
    $scope.nodeSelected = undefined;
    $scope.requisitionGroup.supervisoryNode = undefined;
  };

  function loadSupervisoryNode() {
    if ($scope.requisitionGroup) {
      $scope.nodeSelected = $scope.requisitionGroup.supervisoryNode;
    }
  }
}

RequisitionGroupController.resolve = {
  requisitionGroupData: function ($q, $route, $timeout, RequisitionGroups) {
    if ($route.current.params.id === undefined) return undefined;

    var deferred = $q.defer();
    var requisitionGroupId = $route.current.params.id;

    $timeout(function () {
      RequisitionGroups.get({id: requisitionGroupId}, function (data) {
        deferred.resolve(data.requisitionGroupData);
      }, {});
    }, 100);
    return deferred.promise;
  },

  programs: function ($q, $timeout, Programs) {
    var deferred = $q.defer();

    $timeout(function () {
      Programs.get({type: "pull"}, function (data) {
        deferred.resolve(data.programs);
      }, {});
    }, 100);
    return deferred.promise;
  },

  schedules: function ($q, $timeout, Schedule) {
    var deferred = $q.defer();

    $timeout(function () {
      Schedule.get({}, function (data) {
        deferred.resolve(data.schedules);
      }, {});
    }, 100);
    return deferred.promise;
  }
};

File: modules/openlmis-web/src/main/java/org/openlmis/web/controller/RequisitionGroupController.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 *  This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 *  You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
 */

package org.openlmis.web.controller;

import lombok.NoArgsConstructor;
import org.openlmis.core.domain.Pagination;
import org.openlmis.core.domain.RequisitionGroup;
import org.openlmis.core.domain.RequisitionGroupMember;
import org.openlmis.core.domain.RequisitionGroupProgramSchedule;
import org.openlmis.core.exception.DataException;
import org.openlmis.core.service.RequisitionGroupProgramScheduleService;
import org.openlmis.core.service.RequisitionGroupService;
import org.openlmis.core.service.StaticReferenceDataService;
import org.openlmis.core.web.controller.BaseController;
import org.openlmis.web.form.RequisitionGroupFormDTO;
import org.openlmis.core.web.OpenLmisResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

import static org.openlmis.core.web.OpenLmisResponse.error;
import static org.openlmis.core.web.OpenLmisResponse.success;
import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.web.bind.annotation.RequestMethod.*;

/**
 * This controller handles endpoint related to GET, PUT, POST and other operations on RequisitionGroup.
 */

@Controller
@NoArgsConstructor
public class RequisitionGroupController extends BaseController {

  public static final String SEARCH_PAGE_SIZE = "search.page.size";

  @Autowired
  private RequisitionGroupService requisitionGroupService;

  @Autowired
  private StaticReferenceDataService staticReferenceDataService;

  @Autowired
  private RequisitionGroupProgramScheduleService requisitionGroupProgramScheduleService;

  @RequestMapping(value = "/requisitionGroups", method = GET)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_REQUISITION_GROUP')")
  public ResponseEntity<OpenLmisResponse> search(@RequestParam(value = "searchParam") String searchParam,
                                                 @RequestParam(value = "columnName") String columnName,
                                                 @RequestParam(value = "page", defaultValue = "1") Integer page) {
    Integer pageSize = Integer.parseInt(staticReferenceDataService.getPropertyValue(SEARCH_PAGE_SIZE));
    Pagination pagination = new Pagination(page, pageSize);

    List<RequisitionGroup> requisitionGroupList = requisitionGroupService.search(searchParam, columnName, pagination);
    pagination.setTotalRecords(requisitionGroupService.getTotalRecords(searchParam, columnName));

    ResponseEntity<OpenLmisResponse> response = OpenLmisResponse.response("requisitionGroupList", requisitionGroupList);
    response.getBody().addData("pagination", pagination);
    return response;
  }

  @RequestMapping(value = "/requisitionGroups/{id}", method = GET)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_REQUISITION_GROUP')")
  public ResponseEntity<OpenLmisResponse> getById(@PathVariable(value = "id") Long id) {
    RequisitionGroup requisitionGroup = requisitionGroupService.getBy(id);
    List<RequisitionGroupMember> requisitionGroupMembers = requisitionGroupService.getMembersBy(id);
    List<RequisitionGroupProgramSchedule> requisitionGroupProgramSchedules = requisitionGroupProgramScheduleService.getByRequisitionGroupId(
      id);
    RequisitionGroupFormDTO requisitionGroupFormDTO = new RequisitionGroupFormDTO(requisitionGroup,
      requisitionGroupMembers, requisitionGroupProgramSchedules);
    return OpenLmisResponse.response("requisitionGroupData", requisitionGroupFormDTO);
  }

  @RequestMapping(value = "/requisitionGroups", method = POST, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_REQUISITION_GROUP')")
  public ResponseEntity<OpenLmisResponse> insert(@RequestBody RequisitionGroupFormDTO requisitionGroupFormDTO,
                                                 HttpServletRequest request) {
    try {
      Long userId = loggedInUserId(request);
      requisitionGroupService.saveWithMembersAndSchedules(requisitionGroupFormDTO.getRequisitionGroup(),
        requisitionGroupFormDTO.getRequisitionGroupMembers(),
        requisitionGroupFormDTO.getRequisitionGroupProgramSchedules(), userId);
    } catch (DataException e) {
      return error(e, BAD_REQUEST);
    }

    ResponseEntity<OpenLmisResponse> responseEntity = success(
      messageService.message("message.requisition.group.created.success",
        requisitionGroupFormDTO.getRequisitionGroup().getName()));
    responseEntity.getBody().addData("requisitionGroupId", requisitionGroupFormDTO.getRequisitionGroup().getId());
    return responseEntity;
  }

  @RequestMapping(value = "/requisitionGroups/{id}", method = PUT, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_REQUISITION_GROUP')")
  public ResponseEntity<OpenLmisResponse> update(@PathVariable(value = "id") Long id,
                                                 @RequestBody RequisitionGroupFormDTO requisitionGroupFormDTO,
                                                 HttpServletRequest request) {
    try {
      Long userId = loggedInUserId(request);
      requisitionGroupFormDTO.getRequisitionGroup().setId(id);
      requisitionGroupService.updateWithMembersAndSchedules(requisitionGroupFormDTO.getRequisitionGroup(),
        requisitionGroupFormDTO.getRequisitionGroupMembers(),
        requisitionGroupFormDTO.getRequisitionGroupProgramSchedules(), userId);

    } catch (DataException e) {
      return error(e, BAD_REQUEST);
    }
    ResponseEntity<OpenLmisResponse> responseEntity = success(
      messageService.message("message.requisition.group.updated.success",
        requisitionGroupFormDTO.getRequisitionGroup().getName()));
    responseEntity.getBody().addData("requisitionGroupId", requisitionGroupFormDTO.getRequisitionGroup().getId());
    return responseEntity;
  }

}
